[DataGrid] Make the pinned rows be on top of the no rows overlay

DanailH opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Fixes #9927

MBilalShafi commented on packages/grid/x-data-grid-pro/src/components/DataGridProVirtualScroller.tsx:
```diff
@@ -170,8 +170,8 @@ const VirtualScrollerPinnedRows = styled('div', {
   const boxShadowColor = getBoxShadowColor(theme);
```
I am wondering if it'd be better to not show no rows overlay at all even if there are some pinned rows.

```diff
diff --git a/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx b/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx
index 1b1068045..21183218e 100644
--- a/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx
+++ b/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx
@@ -10,6 +10,7 @@ import { gridExpandedRowCountSelector } from '../../hooks/features/filter/gridFi
 import {
   gridRowCountSelector,
   gridRowsLoadingSelector,
+  gridPinnedRowsCountSelector,
 } from '../../hooks/features/rows/gridRowsSelector';
 import { useGridApiContext } from '../../hooks/utils/useGridApiContext';
 import { useGridRootProps } from '../../hooks/utils/useGridRootProps';
@@ -95,10 +96,11 @@ export function GridOverlays() {
   const rootProps = useGridRootProps();
 
   const totalRowCount = useGridSelector(apiRef, gridRowCountSelector);
+  const pinnedRowsCount = useGridSelector(apiRef, gridPinnedRowsCountSelector);
   const visibleRowCount = useGridSelector(apiRef, gridExpandedRowCountSelector);
   const loading = useGridSelector(apiRef, gridRowsLoadingSelector);
 
-  const showNoRowsOverlay = !loading && totalRowCount === 0;
+  const showNoRowsOverlay = !loading && totalRowCount === 0 && pinnedRowsCount === 0;
   const showNoResultsOverlay = !loading && totalRowCount > 0 && visibleRowCount === 0;
 
   let overlay: React.JSX.Element | null = null;
```

Because:

1. No rows not _actually_ true because we do have some rows (even though they are pinned but there are some rows).
2. Even if we assume that we are talking about the unpinned rows, the styling of no rows overlay isn't consistent as the text is always rendered in the middle, e.g compare it with 4 top pinned rows vs 2 top and two bottom pinned rows.
  - 4 top pinned (No rows overlay rendered but hidden behind pinned rows)
   ![image](https://github.com/mui/mui-x/assets/12609561/240386ec-5d5f-4300-bd09-68f3361da481)
  - 2 top pinned and 2 bottom pinned (shows as expected)
   ![image](https://github.com/mui/mui-x/assets/12609561/5cd4f999-8320-4df5-b4ec-0dba098f619d)

What do you think?

DanailH replied:
That can also be an option. I'm ok with either one, to be honest.

romgrk replied:
We could also clarify further by having `No unpinned rows` in that case.

flaviendelangle commented:
Is it something that is testable?
We had quite a lot of z-index related issues and I feel like everytime we change something we have 75% chance of breaking something else :laughing:

DanailH commented:
> Is it something that is testable? We had quite a lot of z-index related issues and I feel like everytime we change something we have 75% chance of breaking something else 😆

hmm I monkey-tested it with what pop-ups, overlays, and tooltips we have 😄

romgrk commented on packages/grid/x-data-grid-pro/src/components/DataGridProVirtualScroller.tsx:
```diff
@@ -170,8 +170,8 @@ const VirtualScrollerPinnedRows = styled('div', {
     position: 'sticky',
-    // should be above the detail panel
-    zIndex: 3,
+    // should be above the no rows overlay
+    zIndex: 5,
```
I like @yaredtsy's suggestion of adding the message over the unpinned content instead of over the whole content, but I'm not strongly opposed to this.

MBilalShafi replied:
Yes, showing the overlay only over the unpinned content + making the text `No unpinned rows` would be a better XP
Or better, 
```js
const message = pinnedRows.length > 0 ? 'No unpinned rows' : 'No rows'
```

DanailH replied:
This combination is probably the best direction. I'll update the PR accordingly.

DanailH commented:
@romgrk @MBilalShafi I've updated the solution to include your comments. I've moved the `GridOverlays` component to an inner level and added a new `noUnpinnedRowsOverlay` slot component. This is the result

![Screenshot 2023-08-15 at 16 00 08](https://github.com/mui/mui-x/assets/5858539/1c3118e5-9b4a-4cd7-9b5a-6c959809ad40)

![Screenshot 2023-08-15 at 15 59 56](https://github.com/mui/mui-x/assets/5858539/254b3cc8-19db-435b-999d-c99a7fa9727e)

MBilalShafi commented on packages/grid/x-data-grid/src/components/base/GridOverlays.tsx:
```diff
@@ -97,9 +102,11 @@ export function GridOverlays() {
 
   const showNoRowsOverlay = !loading && totalRowCount === 0;
   const showNoResultsOverlay = !loading && totalRowCount > 0 && visibleRowCount === 0;
+  const snowNoUnpinnedRowsOverlay = !loading && totalRowCount === 0 && pinnedRowsCount > 0;
```
_Optional:_ Do we have customization usecases which require a dedicated slot (`noUnpinnedRowsOverlay`) for this use-case. Could we avoid this slot and use this condition (`pinnedRowCount > 0`) in (`noRowsOverlay`) and render different **localeText** based on it?

DanailH replied:
Idk to be honest, I did it like this because it seems to have templates for the different use cases. This does give higher customizability.

cherniavskii replied:
+1 to have a single `noRowsOverlay` slot to make the customization easier

MBilalShafi reviewed (approved):
Looks solid 🎉

romgrk commented on packages/grid/x-data-grid/src/constants/localeTextConstants.ts:
```diff
@@ -3,6 +3,7 @@ import { GridLocaleText } from '../models/api/gridLocaleTextApi';
 export const GRID_DEFAULT_LOCALE_TEXT: GridLocaleText = {
   // Root
   noRowsLabel: 'No rows',
+  noUnpinnedRowsOverlayLabel: 'No unpinned rows found.',
   noResultsOverlayLabel: 'No results found.',
```
Nitpick but the dot at the end of those labels seems a bit off to me. In general our labels don't take final dots because they're not really sentences, I'm wondering why we have one here.

DanailH replied:
Good question, Idk to be honest but it looks more correct punctuation wise 🤔

cherniavskii commented on packages/grid/x-data-grid-pro/src/components/DataGridProVirtualScroller.tsx:
```diff
@@ -443,6 +441,7 @@ const DataGridProVirtualScroller = React.forwardRef<
         </VirtualScrollerPinnedRows>
       ) : null}
       <GridVirtualScrollerContent {...contentProps}>
+        <GridOverlays />
```
This change made the loading overlay move its position on scroll 😅

https://github.com/mui/mui-x/assets/13808724/dc4c4d4b-3c21-47af-bb2c-4aa55c0fdfd0

DanailH replied:
mm a bit of investigation - this one would be tricky not only because of the position of the overlay but because if it is `loading` then the overlay needs to cover all the rows (including the pinned ones)

romgrk replied:
Makes sense, I think leaving the overlay where it was is better then.

DanailH replied:
I opted for a 3rd approach - keep the newly created slot and create an additional component `GridInnerOverlay` that is added deeper in the DOM. The result is no breaking changes to existing functionality and still the issue with the pinned columns is resolved.
If this is not an option then we are back to the simple fix of `z-index: 5` 😅

romgrk replied:
Well the simple fix does seem nice in light of this information :sweat_smile: so I would suggest to go back to that at this point.

cherniavskii commented on packages/grid/x-data-grid-pro/src/components/DataGridProVirtualScroller.tsx:
```diff
@@ -170,8 +170,8 @@ const VirtualScrollerPinnedRows = styled('div', {
-    // should be above the detail panel
-    zIndex: 3,
+    // should be above the no rows overlay
+    zIndex: 5,
```
Should we increase zIndex for loading overlay?

<img width="808" src="https://github.com/mui/mui-x/assets/13808724/9f9da001-e49e-46d6-9e2d-a593946e5462">

DanailH replied:
It's not going to work that way. Both the loading and the no rows overlays use the same container that has the `z-index`

cherniavskii replied:
We can change the zIndex depending on the overlay type, something like this:

```diff
diff --git a/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx b/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx
index 1b1068045..7714e7c14 100644
--- a/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx
+++ b/packages/grid/x-data-grid/src/components/base/GridOverlays.tsx
@@ -21,14 +21,17 @@ const GridOverlayWrapperRoot = styled('div', {
   name: 'MuiDataGrid',
   slot: 'OverlayWrapper',
   overridesResolver: (props, styles) => styles.overlayWrapper,
-})({
+})<{ overlayType: 'loadingOverlay' | string }>(({ overlayType }) => ({
   position: 'sticky', // To stay in place while scrolling
   top: 0,
   left: 0,
   width: 0, // To stay above the content instead of shifting it down
   height: 0, // To stay above the content instead of shifting it down
-  zIndex: 4, // Should be above pinned columns, pinned rows and detail panel
-});
+  zIndex:
+    overlayType === 'loadingOverlay'
+      ? 5 // Should be above pinned columns, pinned rows, and detail panel
+      : 4, // Should be above pinned columns and detail panel
+}));
 
 const GridOverlayWrapperInner = styled('div', {
   name: 'MuiDataGrid',
@@ -49,7 +52,7 @@ const useUtilityClasses = (ownerState: OwnerState) => {
   return composeClasses(slots, getDataGridUtilityClass, classes);
 };
 
-function GridOverlayWrapper(props: React.PropsWithChildren<{}>) {
+function GridOverlayWrapper(props: React.PropsWithChildren<{ overlayType: string }>) {
   const apiRef = useGridApiContext();
   const rootProps = useGridRootProps();
 
@@ -77,7 +80,7 @@ function GridOverlayWrapper(props: React.PropsWithChildren<{}>) {
   }
 
   return (
-    <GridOverlayWrapperRoot className={clsx(classes.root)}>
+    <GridOverlayWrapperRoot className={clsx(classes.root)} overlayType={overlayType}>
       <GridOverlayWrapperInner
         className={clsx(classes.inner)}
         style={{
@@ -102,16 +105,20 @@ export function GridOverlays() {
   const showNoResultsOverlay = !loading && totalRowCount > 0 && visibleRowCount === 0;
 
   let overlay: React.JSX.Element | null = null;
+  let overlayType = '';
 
   if (showNoRowsOverlay) {
+    overlayType = 'noRowsOverlay';
     overlay = <rootProps.slots.noRowsOverlay {...rootProps.slotProps?.noRowsOverlay} />;
   }
 
   if (showNoResultsOverlay) {
+    overlayType = 'noResultsOverlay';
     overlay = <rootProps.slots.noResultsOverlay {...rootProps.slotProps?.noResultsOverlay} />;
   }
 
   if (loading) {
+    overlayType = 'loadingOverlay';
     overlay = <rootProps.slots.loadingOverlay {...rootProps.slotProps?.loadingOverlay} />;
   }
 
@@ -119,5 +126,5 @@ export function GridOverlays() {
     return null;
   }
 
-  return <GridOverlayWrapper>{overlay}</GridOverlayWrapper>;
+  return <GridOverlayWrapper overlayType={overlayType}>{overlay}</GridOverlayWrapper>;
 }

```

DanailH replied:
Got it. Updated

romgrk replied:
Personal preference but I don't like too much using strings as constants. I generally do something like this:

```typescript
enum OverlayType {
  NoRows,
  NoResults,
  Loading,
}
```

DanailH replied:
Yeah, I feel you. We had this argument so many times that idk to be honest which one is the better way. I honestly would opt for no strings but I know we had an unwritten rule that if something is used only once then we use strings (at least this was in the past).

romgrk replied:
I prefer enums over unions because enums (without default values) use integers instead of strings. In many cases (this one included) it doesn't make much of a difference, but string comparisons can bring the performance down if used carelessly. [Here](https://github.com/json5/json5/pull/278) is an admittedly very specific use-case where string usage causes a 2x slowdown.

I also find enums easier to reason about, they feel less like a hack imho.

cherniavskii replied:
> https://github.com/json5/json5/pull/278 is an admittedly very specific use-case where string usage causes a 2x slowdown.

Nice, great work! I'm surprised it hasn't been merged yet 😅

> I also find enums easier to reason about, they feel less like a hack IMHO.

I'm fine with using both internally depending on the use case and personal preference 👍🏻

When it comes to the public API and DX, how do you feel about this?

```tsx
import { DataGrid } from '@mui/x-data-grid';
// string union
<DataGrid filterMode="server" />
```

```tsx
import { DataGrid, GridFilterMode } from '@mui/x-data-grid';
// enum
<DataGrid filterMode={GridFilterMode.Server} />
```

I definitely prefer string union, because:
1. No need to import any additional types
2. Autocomplete

cherniavskii replied:
@MBilalShafi feel free to chime in 🙂

MBilalShafi replied:
Some really good arguments made and awesome work with the `parse()` performance PR Rom 🚀 

I kind of agree with preferring the unions, at least for the public API as the users will not have to learn what's a specific enum is called and import it, filterMode="server" is a good example.

But when there's already a learning curve, like with the overlay names, which we define and if customization is needed users will have to know the names anyways. In such cases, enums could make more sense as it looks more readable on top of the slight performance benefit it gives by using integers instead of strings.

romgrk replied:
> Nice, great work! I'm surprised it hasn't been merged yet 😅

I had a feeling the maintainer wouldn't merge. It's good for performance, bad for readability. Did it mostly for fun though :sweat_smile: 

For the public API, I'll agree that unions are more ergonomic. JSX props feel more natural with a string than with an enum.

Internally, for the large majority of cases it won't matter, but imho it would be a good habit to use enums because in the few cases where it matters, we'll already be using the efficient solution.
